web/pdf_viewer.js

```diff
@@ -2256,13 +2297,54 @@ class PDFViewer {
     if (!this.pdfDocument) {
       return;
     }
-    this.#annotationEditorMode = mode;
-    this.eventBus.dispatch("annotationeditormodechanged", {
-      source: this,
-      mode,
-    });
 
-    this.#annotationEditorUIManager.updateMode(mode, editId, isFromKeyboard);
+    const { eventBus } = this;
+    const updater = () => {
+      if (this.#onPageRenderedCallback) {
+        eventBus._off("pagerendered", this.#onPageRenderedCallback);
+        this.#onPageRenderedCallback = null;
+      }
+      if (this.#switchAnnotationEditorModeTimeoutId) {
+        clearTimeout(this.#switchAnnotationEditorModeTimeoutId);
+        this.#switchAnnotationEditorModeTimeoutId = 0;
+      }
+      this.#annotationEditorMode = mode;
+      eventBus.dispatch("annotationeditormodechanged", {
+        source: this,
+        mode,
+      });
+      this.#annotationEditorUIManager.updateMode(mode, editId, isFromKeyboard);
+    };
+
+    if (
+      mode === AnnotationEditorType.NONE ||
+      this.#annotationEditorMode === AnnotationEditorType.NONE
+    ) {
+      const isEditing = mode !== AnnotationEditorType.NONE;
+      if (!isEditing) {
+        this.pdfDocument.annotationStorage.resetModifiedIds();
+      }
+      for (const pageView of this._pages) {
+        pageView.toggleEditingMode(isEditing);
+      }
+      const idsToRefresh = this.#switchToEditAnnotationMode();
+      if (isEditing && editId && idsToRefresh) {
+        // We're editing an existing annotation so we must switch to editing
+        // mode when the rendering is done.
+        const { signal } = this.#eventAbortController;
+        this.#onPageRenderedCallback = ({ source: { id } }) => {
```

Comment:
Relying on the `source` seems unnecessary; can't we instead use one of the "normal" parameters since that one should contain the same data?
```suggestion
        this.#onPageRenderedCallback = ({ pageNumber }) => {
```